Update Python Editing docs with more code actions (#7265)
microsoft/vscode-docs: docs/python/editing.md

## Quick Fixes

### Add import

> The add import Quick Fix when using Pylance allows you to quickly complete import statements. First, begin by typing a package name within the editor. You will notice a Code Action is available to automatically complete the line of source code (as long as you have the module installed within the environment). Hover over the text (marked with a squiggle) and then select the Code Action light bulb when it appears. You can then select from a list of potential imports.

Reviewer: ```suggestion
When using Pylance, the add import Quick Fix enables you to quickly complete import statements for modules that are installed in your environment. As you start typing a package name in the editor, a Code Action is available to automatically complete the line of source code. Hover over the text (marked with a squiggle) and select the Code Action light bulb. You can then select from the list of potential imports.
```

![Add import code action](images/editing/quickFix.png)

This Code Action also recognizes some of the popular abbreviations for the following common Python packages: `numpy` as np, `tensorflow` as tf, `pandas` as pd, `matplotlib.pyplot` as plt, `matplotlib` as mpl, `math` as m, `scipi.io` as spio, and `scipy` as sp, `panel` as pn, and `holoviews` as hv.

![Common package abbreviations](images/editing/packageAbbreviations.gif)

The import suggestions list displays the top 3 high-confidence import options, prioritized based on: most recently used imports, symbols from the same module, symbols from the standard library, symbols from user modules, symbols from third-party packages, and finally sorting by module and symbol name.

### Search for additional import matches

> The add import Quick Fix only shows 3 high-confidence import options by default. In the case where they aren't what you are looking for, Pylance has a **Search for additional import matches** Quick Fix for missing import errors, which also displays a quick pick menu that allows you to search for import options that prefix-match the missing import symbol.

Reviewer: ```suggestion
By default, the add import Quick Fix only shows 3 high-confidence import options. If they don't list what you are looking for, you can use the Pylance **Search for additional import matches** Quick Fix for missing import errors. This Quick Fix displays a quick pick menu that enables you to search for import options that prefix-match the missing import symbol.
```

![Search for additional import matches Code Action](images/editing/search-imports-code-action.gif)

### Change spelling

> The **Change spelling** Quick Fix is displayed by Pylance on unresolved variable or missing imports diagnostics when they are likely caused by typos. This Code Action suggests the correct spelling of the symbol based on the closest matches found in the workspace.

Reviewer: ```suggestion
Pylance displays the **Change spelling** Quick Fix on unresolved variables or missing imports diagnostics when they are likely caused by typos. This Code Action suggests the correct spelling of the symbol, based on the closest matches found in the workspace.
```

![Change spelling code action on missing import due to a typo](images/editing/change-spelling-code-action.gif)

> > **Note**: Symbol aliases defined in user code (for example,  the `bar` alias in `import foo as bar`) are not supported by these Quick Fixes for performance reasons. Just like with auto imports, only top-level symbols are suggested by default, but you can customize this behavior for specific packages through the `python.analysis.packageIndexDepths` setting.

Author: ```suggestion
> **Note**: Symbol aliases defined in user code (for example,  the `bar` alias in `import foo as bar`) are not supported by these Quick Fixes for performance reasons. Just like with auto imports, only top-level symbols -- those not nested within classes or functions -- are suggested by default, but you can customize this behavior for specific packages through the `python.analysis.packageIndexDepths` setting.
```

> > **Note**: Symbol aliases defined in user code (e.g.  `bar` on `import foo as bar`) are not supported by these Quick Fixes for performance reasons. Just like with auto imports, only top-level symbols are suggested by default, but you can customize this behavior for specific packages through the `python.analysis.packageIndexDepths` setting.

Reviewer: ```suggestion
> **Note**: Symbol aliases defined in user code (for example,  the `bar` alias in `import foo as bar`) are not supported by these Quick Fixes for performance reasons. Just like with auto imports, only top-level symbols are suggested by default, but you can customize this behavior for specific packages through the `python.analysis.packageIndexDepths` setting.
```

> > **Note**: Symbol aliases defined in user code (e.g.  `bar` on `import foo as bar`) are not supported by these Quick Fixes for performance reasons. Just like with auto imports, only top-level symbols are suggested by default, but you can customize this behavior for specific packages through the `python.analysis.packageIndexDepths` setting.

Reviewer: Is the concept of top-level symbols known by all users?

## Refactorings

> The Python extension adds the following refactoring functionalities via the Pylance extension: **Extract Variable**, **Extract Method**, **Rename Module**, **Move Symbol** and **Implement All Inherited Abstract Classes**. It also supports extensions that implement additional refactoring features, such as **Sort Imports**.

Reviewer 2: It might be nice to provide a link somewhere in the last sentence to the ### Sort Imports section section later in this doc that talks about the extensions that provide this Code Action.

Author: in this case this paragraph is just mentioning all sub sections in the refactoring section. Once the page is published we can revisit this and add al links in case we think it'll make it easier to navigate between them :)

> The Python extension adds the following refactoring functionalities via the Pylance extension: **Extract Variable**, **Extract Method**, **Rename Module**, **Move Symbol** and **Implement All Inherited Abstract Classes**. It also supports extensions that implement additional refactoring features such as **Sort Imports**.

Reviewer: ```suggestion
The Python extension adds the following refactoring functionalities via the Pylance extension: **Extract Variable**, **Extract Method**, **Rename Module**, **Move Symbol** and **Implement All Inherited Abstract Classes**. It also supports extensions that implement additional refactoring features, such as **Sort Imports**.
```

## Quick Fixes

### Add Missing Imports

The add imports Quick Fix when using Pylance allows you to quickly complete import statements. First, begin by typing a package name within the editor. You will notice a Code Action is available to automatically complete the line of source code (as long as you have the module installed within the environment). Hover over the text (marked with a squiggle) and then select the Code Action light bulb when it appears. You can then select from a list of potential imports.

![Add import code action](images/editing/quickFix.png)

This Code Action also recognizes some of the popular abbreviations for the following common Python packages: `numpy` as np, `tensorflow` as tf, `pandas` as pd, `matplotlib.pyplot` as plt, `matplotlib` as mpl, `math` as m, `scipi.io` as spio, and `scipy` as sp, `panel` as pn, and `holoviews` as hv.

![Common package abbreviations](images/editing/packageAbbreviations.gif)

> The import suggestions list displays  the top 3 high-confidence import options, prioritized based on: most recently used imports, symbols from the same module, symbols from the standard library, symbols from user modules, symbols from third-party packages, and finally sorting by module and symbol name.  In the case where the 3 high-confidence import options aren’t what you are looking for, Pylance has a **Search for additional import matches** Code Action, which displays a quick pick menu that allows you to search for import options that prefix-match the missing import symbol.

Reviewer 3: Nitpick: I'm not sure if this actually affects markdown rendering.

```suggestion
The import suggestions list displays the top 3 high-confidence import options, prioritized based on: most recently used imports, symbols from the same module, symbols from the standard library, symbols from user modules, symbols from third-party packages, and finally sorting by module and symbol name.  In the case where the 3 high-confidence import options aren’t what you are looking for, Pylance has a **Search for additional import matches** Code Action, which displays a quick pick menu that allows you to search for import options that prefix-match the missing import symbol.
```

![Search for additional import matches Code Action](images/editing/search-imports-code-action.gif)

Pylance also offers a **Change spelling** Code Action, which displays import suggestions for missing imports due to typos.

![Change spelling code action on missing import due to a typo](images/editing/change-spelling-code-action.gif)

> Just like with auto imports, only top-levels symbols are suggested by default. You can customize this behavior through the `python.analysis.packageIndexDepths` setting. Note that user symbol aliases (e.g. `import foo as bar`) are not supported by these Code Actions.

Reviewer 3: >  Note that user symbol aliases (e.g. `import foo as bar`) are not supported by these Code Actions.

I'm not sure what this means. Is it referring to the popular abbreviation list (ex. pandas as pd) mentioned above? Or is it saying that it doesn't support adding a custom alias to the import statement it generates? Or something else?

Reviewer 4: what it trying to say is when there is

```
module/__init__.py
from module.impl import symbol as symbol

module/impl.py
class symbol: ...
```

when module is 3rd party library, we will show `symbol` from `module/__init__` in `add/auto import`, but when module is user module/file, we will show `symbol` from `module/impl`.

in another word, we don't support `alias symbol` from `user file/module` in `add/auto import`

Reviewer 3: I think the current phrasing is going to confuse users, but I'm not sure how to make it more clear. Maybe this isn't important enough to mention here?

I'm also wondering if this is the right place to mention it, since this paragraph is about indexing depth but that isn't an issue with user code, is it?

Reviewer 4: ya, that's for 3rd party packages. 

but also, I believe we want to add somewhere that we currently don't support alias symbol from user files since that question comes up again and again from users.

Author: I think the confusion may be because it was brought up after the indexing depth setting, so it felt like they were talking about similar issues when in fact the paragraph is just acknowledging some limitations? Coming from that assumption, I just flipped the order of the sentence and put it all under a "note" block to try to clarify that

### Implement All Inherited Abstract Classes

> In Python, abstract classes serve as "blueprints" for other classes and help build modular, reusable code by promoting clear structure and requirements for subclasses to adhere to. To define an abstract class in Python, you can create a class that inherits from the `ABC` class in the `abc` module, and annotate its methods with the `@abstractmethod` decorator. Then, you can create new classes that inherit from this abstract class, and define an implementation for the base methods.

Reviewer 5: I added this code action as refactoring since my understanding is that quick fix is to get rid of the diagnostics. [Another enhancement](https://github.com/microsoft/pylance-release/issues/5757) that Heejae did to fix the `reportAbstractUsage` diagnostic is a quick fix.

Author: very good point, thanks for bringing this up! I moved this to the refactoring section, and once the quick fix lands on stable I can add it back to the quick fix section :)

> In Python, abstract classes serve as "blueprints" for other classes and help build modular, reusable code by promoting clear structure and requirements for subclasses to adhere to. To define an abstract class in Python, you can create a class that inherits from the `ABC` class in the `abc` module, and annotate its methods with the `@abstractmethod` decorator. Then, you can create new classes that inherit from this abstract class, and define an implementation for the base methods.

Reviewer 3: > and annotate its methods with the `@abstractmethod` decorator.

Should we also mention `@abstractclassmethod`, `@abstractstaticmethod`, and `@abstractproperty`? And/or point to Python docs on abstract classes for more details?

Reviewer 5: Those decorators are deprecated since 3.3, not sure if worth mentioning it.

Reviewer 3: Good point, thanks.

![Renaming a module](images/editing/refactorRenameModule.gif)

### Move Symbol

> The Pylance extension offers two Code Actions to simplify the process of moving symbols to different files: **Move symbol to...**, which displays a file picker so you can select the destination file for the symbol to be moved to,  and **Move symbol to new file**, which creates a new file with the symbol name, located on the same directory as the source file where the Code Action was invoked.

Reviewer: ```suggestion
The Pylance extension offers two Code Actions to simplify the process of moving symbols to different files: 

* **Move symbol to...**: displays a file picker to select the destination file for the symbol to be moved to.
* **Move symbol to new file**: creates a new file with the symbol name, located in the same directory as the source file where the Code Action was invoked.
```

You can access these Code Actions by hovering over the symbol you want to move, then selecting the light bulb that appears next to the desired action. Alternatively, you can right-click on the symbol and select **Refactor...** from the context menu.

![Move Symbol refactoring options](images/editing/move-symbol.gif)

### Implement All Inherited Abstract Classes

In Python, abstract classes serve as "blueprints" for other classes and help build modular, reusable code by promoting clear structure and requirements for subclasses to adhere to. To define an abstract class in Python, you can create a class that inherits from the `ABC` class in the `abc` module, and annotate its methods with the `@abstractmethod` decorator. Then, you can create new classes that inherit from this abstract class, and define an implementation for the base methods.

> Pylance offers a code action to simplify the process of creating these classes. When defining a new class that inherits from an abstract one, you can now use the **Implement all inherited abstract classes** Code Action to automatically implement all abstract methods and properties from the parent class:

Reviewer: ```suggestion
Pylance offers a code action to simplify the process of creating these classes. When you define a new class that inherits from an abstract one, you can now use the **Implement all inherited abstract classes** Code Action to automatically implement all abstract methods and properties from the parent class:
```

* **Move symbol to...**: displays a file picker to select the destination file for the symbol to be moved to.
* **Move symbol to new file**: creates a new file with the symbol name, located in the same directory as the source file where the Code Action was invoked.

You can access these Code Actions by hovering over the symbol you want to move, then selecting the light bulb that appears next to the desired action. Alternatively, you can right-click on the symbol and select **Refactor...** from the context menu.

![Move Symbol refactoring options](images/editing/move-symbol.gif)

### Implement All Inherited Abstract Classes

In Python, abstract classes serve as "blueprints" for other classes and help build modular, reusable code by promoting clear structure and requirements for subclasses to adhere to. To define an abstract class in Python, you can create a class that inherits from the `ABC` class in the `abc` module, and annotate its methods with the `@abstractmethod` decorator. Then, you can create new classes that inherit from this abstract class, and define an implementation for the base methods.

> Pylance offers a code action to simplify the process of creating these classes. When you define a new class that inherits from an abstract one, you can now use the **Implement all inherited abstract classes** Code Action to automatically implement all abstract methods and properties from the parent class:

Reviewer 2: ```suggestion
Pylance offers a Code Action to simplify the process of creating these classes. When you define a new class that inherits from an abstract one, you can now use the **Implement all inherited abstract classes** Code Action to automatically implement all abstract methods and properties from the parent class:
```

![Renaming a module](images/editing/refactorRenameModule.gif)

### Move Symbol

> The Pylance extension offers two Code Actions to simplify the process of moving symbols to different files: **Move symbol to...**, which displays a file picker so you can select the destination file for the symbol to be moved to,  and **Move symbol**, which creates a new file with the symbol name, located on the same directory as the source file where the Code Action was invoked.

Reviewer 3: ```suggestion
The Pylance extension offers two Code Actions to simplify the process of moving symbols to different files: **Move symbol to...**, which displays a file picker so you can select the destination file for the symbol to be moved to,  and **Move symbol to new file**, which creates a new file with the symbol name, located on the same directory as the source file where the Code Action was invoked.
```

> You can access this Code Action by hovering over the symbol you want to move, then selecting the light bulb that appears next to it. Alternatively, you can right-click on the symbol and select **Refactor...** from the context menu.

Reviewer 3: ```suggestion
You can access these Code Actions by hovering over the symbol you want to move, then selecting the light bulb that appears next to the desired action. Alternatively, you can right-click on the symbol and select **Refactor...** from the context menu.
```
